### ISSUE TITLE
Some way to configure global settings

### ISSUE THREAD

**Issue by jjrv** [phosphorjs/phosphor-ui#17]
There could be some way to configure settings like [`DETACH_THRESHOLD`](https://github.com/phosphorjs/phosphor-ui/blob/e751b0ee39d837d91c1a513d1bb50f0badb9ee03/src/tabbar.ts#L125). I've changed to dockpanel to work so that dragging tabs on top of other tabbars docks them as new tabs (same as dragging to the center of another panel used to do). I'm also planning to make dragging to the center of another panel float the tab as a movable window instead of docking it.

I'm guessing you won't take PRs for the new features without a way to turn them on/off since someone must prefer the current functionality instead. So if there was an overall way to pass options to Phosphor widget JavaScript code, I could add flags for these things and them maybe get the PRs to you.

**Comment by sccolbert** [comment]
Yep. Making this configurable is certainly something I'm open to.

I responded to your other issue directly. I'm definitely interested about make the dock areas better/easier to use.

**Comment by jjrv** [comment]
So as an idea, we could use the new module augmentation feature in TypeScript 1.8. There could be a top-level PhosphorJS options interface, and each widget when imported, would define an interface for its own options and augment the top-level interface to hold a member for configuring that widget type.

When creating a widget, the top-level options could be passed to it. It would pass that object on to any other widgets it creates. That means I can for example import the dockpanel and tabbar. That gives me:

```TypeScript
interface IPhosphorOptions {
  dockPanelOptions: IDockPanelOptions;
  tabBarOptions: ITabBarOptions;
}

interface ITabBarOptions {
  /** The start drag distance threshold. */
  dragThreshold: number;

  // etc.
}
```

That would allow me to create 2 dockpanels with different options for tabbars inside each one, if that was necessary for some reason (for example combining 3rd party widgets from different sources, which require particular settings for their children to look correct). All I'd need is to give them different `IPhosphorOptions` objects.

Probably the CSS class names (like the current `const TAB_BAR_CLASS = 'p-TabBar';`) should be configurable inside the options objects as well.

Each widget will have a fully typed view of its own settings and the settings of any other widgets it imports, and is able to pass on settings it doesn't understand while preserving types. If 3rd parties create new widgets, they can use the same principles to extend `IPhosphorOptions` and their options will be carried from dockpanels to child widgets. Everything stays fully typed, and types for all imported widgets (and no others) are seen by the compiler.

**Comment by jjrv** [comment]
PR #20 shows how it would work. Note [this module augmentation](https://github.com/phosphorjs/phosphor-ui/blob/fe4124420ae888264fe964d90be869021e31ebeb/src/tabbar.ts#L56-L60) where the tab bar widget injects its configuration object type into a top-level configuration type.
